diff --git a/src/input/mod.rs b/src/input/mod.rs
index ab31df94..1bbee77a 100644
--- a/src/input/mod.rs
+++ b/src/input/mod.rs
@@ -2845,7 +2845,7 @@ impl State {
                     self.niri
                         .layout
                         .view_offset_gesture_begin(&output, Some(ws_idx), false);
-                    let grab = SpatialMovementGrab::new(start_data, output, ws_id, true);
+                    let grab = SpatialMovementGrab::new(start_data, event.time_msec(), output, ws_id, true);
                     pointer.set_grab(self, grab, serial, Focus::Clear);
                     self.niri
                         .cursor_manager
@@ -2858,8 +2858,7 @@ impl State {
             }
 
             if button == Some(MouseButton::Middle) && !pointer.is_grabbed() {
-                let mod_down = modifiers_from_state(mods).contains(mod_key.to_modifiers());
-                if mod_down {
+                if true {
                     let output_ws = if is_overview_open {
                         self.niri.workspace_under_cursor(true)
                     } else {
@@ -2882,7 +2881,7 @@ impl State {
                             button: button_code,
                             location,
                         };
-                        let grab = SpatialMovementGrab::new(start_data, output, ws_id, false);
+                        let grab = SpatialMovementGrab::new(start_data, event.time_msec(), output, ws_id, false);
                         pointer.set_grab(self, grab, serial, Focus::Clear);
                         self.niri
                             .cursor_manager
@@ -3078,6 +3077,28 @@ impl State {
             },
         );
         pointer.frame(self);
+
+        if let Some((button, time_press, time_release)) = self.niri.queue_synthesized_click.take() {
+            pointer.button(
+                self,
+                &ButtonEvent {
+                    button,
+                    state: ButtonState::Pressed,
+                    serial: SERIAL_COUNTER.next_serial(),
+                    time: time_press,
+                },
+            );
+            pointer.button(
+                self,
+                &ButtonEvent {
+                    button,
+                    state: ButtonState::Released,
+                    serial: SERIAL_COUNTER.next_serial(),
+                    time: time_release,
+                },
+            );
+            pointer.frame(self);
+        }
     }
 
     fn on_pointer_axis<I: InputBackend>(&mut self, event: I::PointerAxisEvent) {
diff --git a/src/input/spatial_movement_grab.rs b/src/input/spatial_movement_grab.rs
index 546929a0..40595ac0 100644
--- a/src/input/spatial_movement_grab.rs
+++ b/src/input/spatial_movement_grab.rs
@@ -16,6 +16,7 @@ use crate::utils::get_monotonic_time;
 
 pub struct SpatialMovementGrab {
     start_data: PointerGrabStartData<State>,
+    start_time: u32,
     last_location: Point<f64, Logical>,
     output: Output,
     workspace_id: WorkspaceId,
@@ -37,6 +38,7 @@ enum GestureState {
 impl SpatialMovementGrab {
     pub fn new(
         start_data: PointerGrabStartData<State>,
+        start_time: u32,
         output: Output,
         workspace_id: WorkspaceId,
         is_view_offset: bool,
@@ -51,6 +53,7 @@ impl SpatialMovementGrab {
         Self {
             last_location: location,
             start_data,
+            start_time,
             output,
             workspace_id,
             gesture,
@@ -186,8 +189,17 @@ impl PointerGrab<State> for SpatialMovementGrab {
         handle.button(data, event);
 
         if handle.current_pressed().is_empty() {
+            let synthesize = self.gesture == GestureState::Recognizing;
+            let button = self.start_data.button;
+
             // No more buttons are pressed, release the grab.
             handle.unset_grab(self, data, event.serial, event.time, true);
+
+            if synthesize {
+                let time_press = self.start_time;
+                let time_release = event.time;
+                data.niri.queue_synthesized_click = Some((button, time_press, time_release));
+            }
         }
     }
 
diff --git a/src/niri.rs b/src/niri.rs
index 414f702d..00ddd0a9 100644
--- a/src/niri.rs
+++ b/src/niri.rs
@@ -413,6 +413,8 @@ pub struct Niri {
 
     #[cfg(feature = "xdp-gnome-screencast")]
     pub casting: Screencasting,
+
+    pub queue_synthesized_click: Option<(u32, u32, u32)>,
 }
 
 #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
@@ -2567,6 +2569,8 @@ impl Niri {
 
             #[cfg(feature = "xdp-gnome-screencast")]
             casting: screencasting,
+
+            queue_synthesized_click: None,
         };
 
         niri.reset_pointer_inactivity_timer();
